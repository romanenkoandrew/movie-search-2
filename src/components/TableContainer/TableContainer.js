import React from 'react';
import { useLocation } from 'react-router-dom';
import PropTypes from 'prop-types';
import {
  ProfileOutlined,
  ProfileFilled,
  StarOutlined,
  StarFilled,
  PlaySquareOutlined,
  PlaySquareFilled,
  RestOutlined,
  RestFilled,
  DeleteOutlined,
} from '@ant-design/icons';
import { Table, Space, Tooltip, Button, Typography, Row } from 'antd';
import styles from './TableContainer.styles';

const { Title } = Typography;

const columns = [
  {
    title: 'Poster',
    dataIndex: 'Poster',
    key: 'Poster',
    width: 100,
    render: (dataIndex) => {
      return <div css={styles.poster(dataIndex)}></div>;
    },
  },
  {
    title: 'Title',
    dataIndex: 'Title',
    key: 'Title',
  },
  {
    title: 'Year',
    dataIndex: 'Year',
    key: 'Year',
  },
  {
    title: 'Actions',
    key: 'actions',
    render: () => (
      <Space size='small'>
        <Tooltip title='Add to watch list'>
          <Button size='small' shape='circle' icon={<ProfileOutlined />} />
        </Tooltip>
        <Tooltip title='Add to viewed'>
          <Button size='small' shape='circle' icon={<PlaySquareOutlined />} />
        </Tooltip>
        <Tooltip title='Add to favorite'>
          <Button size='small' shape='circle' icon={<StarOutlined />} />
        </Tooltip>
        <Tooltip title='Add to black list'>
          <Button size='small' shape='circle' icon={<RestOutlined />} />
        </Tooltip>
        <Tooltip title='Delete'>
          <Button size='small' shape='circle' icon={<DeleteOutlined />} />
        </Tooltip>
      </Space>
    ),
  },
];

const TableContainer = (props) => {
  const {
    movieTitles,
    watchList,
    viewed,
    favorite,
    blackList,
    isLoading,
    searchBlockValues,
    getAllTitles,
    currentPageIncrement,
    currentPageDecrement,
    currentPage,
    totalResults,
  } = props;
  const { pathname } = useLocation();
  const dataForTable = () => {
    switch (pathname) {
      case '/watchlist': {
        return watchList;
      }
      case '/viewed': {
        return viewed;
      }
      case '/favorite': {
        return favorite;
      }
      case '/blacklist': {
        return blackList;
      }
      default: {
        return movieTitles;
      }
    }
  };
  const titleTable = () => {
    switch (pathname) {
      case '/watchlist': {
        return 'Watch List';
      }
      case '/viewed': {
        return 'Viewed';
      }
      case '/favorite': {
        return 'Favorite';
      }
      case '/blacklist': {
        return 'Black List';
      }
      default: {
        return;
      }
    }
  };
  const data = dataForTable();

  const numberValuesOnPage = 10;
  const { search, type } = searchBlockValues;
  const nextPage = () => {
    if (search) {
      currentPageIncrement({ currentPage });
      getAllTitles({ search: search, type: type });
    }
  };
  const previousPage = () => {
    if (search) {
      currentPageDecrement({ currentPage });
      getAllTitles({ search: search, type: type });
    }
  };

  const diffTable = () => {
    if (pathname !== '/') {
      return (
        <Table
          css={styles.table}
          bordered
          columns={columns}
          dataSource={data}
          scroll={{ x: true }}
          loading={isLoading}
          onRow={(record) => {
            return {
              onClick: (event) => console.log(record, event),
              onDoubleClick: () => console.log(record),
            };
          }}
        />
      );
    } else
      return (
        <>
          <Table
            css={styles.table}
            bordered
            columns={columns}
            dataSource={data}
            scroll={{ x: true }}
            loading={isLoading}
            pagination={false}
            onRow={(record) => {
              return {
                onClick: (event) => console.log(record, event),
                onDoubleClick: () => console.log(record),
              };
            }}
          />
          <Row justify='end'>
            <Button
              css={styles.button}
              onClick={previousPage}
              disabled={currentPage === 1}
            >
              &lt; Previous
            </Button>
            <Button
              css={styles.button}
              onClick={nextPage}
              disabled={
                !search || currentPage > totalResults / numberValuesOnPage
              }
            >
              Next &gt;
            </Button>
          </Row>
        </>
      );
  };

  return (
    <>
      <Title css={styles.titleTable} level={2}>
        {titleTable()}
      </Title>
      {diffTable()}
    </>
  );
};
export default TableContainer;

TableContainer.propTypes = {
  movieTitles: PropTypes.array.isRequired,
  watchList: PropTypes.array.isRequired,
  viewed: PropTypes.array.isRequired,
  favorite: PropTypes.array.isRequired,
  blackList: PropTypes.array.isRequired,
  isLoading: PropTypes.bool.isRequired,
  getAllTitles: PropTypes.func.isRequired,
  currentPageIncrement: PropTypes.func.isRequired,
  currentPageDecrement: PropTypes.func.isRequired,
  searchBlockValues: PropTypes.object.isRequired,
  currentPage: PropTypes.number.isRequired,
  totalResults: PropTypes.number.isRequired,
};
