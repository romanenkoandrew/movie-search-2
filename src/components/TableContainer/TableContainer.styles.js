import { css } from '@emotion/core';

export default {
  poster: (image) => css`
    height: 5rem;
    width: 5rem;
    background-image: url(${image});
    background-size: cover;
  `,
  table: () => css`
    .ant-table-thead > tr > th,
    .ant-table-cell {
      text-align: center;
    }
  `,
  titleTable: () => css`
    text-align: center;
  `,
  button: () => css`
      width: 7rem;
      margin-right: 0.5rem;
      margin-top: 0.5rem;
    }
  `,
};
