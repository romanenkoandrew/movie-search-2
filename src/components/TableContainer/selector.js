import {
  movieTitlesSelector,
  watchListSelector,
  viewedSelector,
  favoriteSelector,
  blackListSelector,
  isLoadingSelector,
  currentPageSelector,
  totalResultsSelector,
} from 'selectors';
import { isOpenModalSelector, aboutMovieSelector } from 'selectors/modalWindow';
import { createStructuredSelector } from 'reselect';

export default createStructuredSelector({
  movieTitles: movieTitlesSelector,
  watchList: watchListSelector,
  viewed: viewedSelector,
  favorite: favoriteSelector,
  blackList: blackListSelector,
  isLoading: isLoadingSelector,
  currentPage: currentPageSelector,
  totalResults: totalResultsSelector,
  isOpenModal: isOpenModalSelector,
  aboutMovie: aboutMovieSelector,
});
