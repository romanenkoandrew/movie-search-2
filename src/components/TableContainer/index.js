import TableContainer from './TableContainer';
import {
  getAllTitles,
  getTitleByID,
  currentPageIncrement,
  currentPageDecrement,
  toggleModal,
} from 'actions';
import { connect } from 'react-redux';
import selector from './selector';

const mapDispatchToProps = {
  getAllTitles,
  getTitleByID,
  currentPageIncrement,
  currentPageDecrement,
  toggleModal,
};

export default connect(selector, mapDispatchToProps)(TableContainer);
