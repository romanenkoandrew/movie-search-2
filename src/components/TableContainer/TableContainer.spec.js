import React from 'react';
import { shallow } from 'enzyme';
import TableContainer from './TableContainer';

jest.mock('react-router-dom', () => ({
  ...jest.requireActual('react-router-dom'),
  useLocation: () => ({
    pathname: 'localhost:3000/',
  }),
}));
const initialState = {
  movieTitles: [],
  watchList: [],
  viewed: [],
  favorite: [],
  blackList: [],
  isLoading: false,
  getAllTitles: () => {},
  getTitleByID: () => {},
  currentPageIncrement: () => {},
  currentPageDecrement: () => {},
  searchBlockValues: {},
  currentPage: 1,
  totalResults: '100',
  isOpenModal: false,
  aboutMovie: {},
  toggleModal: () => {},
  upgradeWatchList: () => {},
  upgradeViewedList: () => {},
  upgradeFavoriteList: () => {},
  upgradeBlackList: () => {},
};
const setUp = () => shallow(<TableContainer {...initialState} />);

describe('TableContainer component', () => {
  let component;
  let instance;
  beforeEach(() => {
    component = setUp();
    instance = component.instance();
  });
  it('should render TableContainer component', () => {
    expect(component).toMatchSnapshot();
  });
});
