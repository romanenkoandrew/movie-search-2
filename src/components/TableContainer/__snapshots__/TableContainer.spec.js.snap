// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`TableContainer component should render TableContainer component 1`] = `
<Fragment>
  <EmotionCssPropInternal
    __EMOTION_LABEL_PLEASE_DO_NOT_USE__="TableContainer"
    __EMOTION_TYPE_PLEASE_DO_NOT_USE__={[Function]}
    css={[Function]}
    level={2}
  />
  <EmotionCssPropInternal
    __EMOTION_TYPE_PLEASE_DO_NOT_USE__={[Function]}
    bordered={true}
    columns={
      Array [
        Object {
          "dataIndex": "Poster",
          "key": "Poster",
          "render": [Function],
          "title": "Poster",
          "width": 100,
        },
        Object {
          "dataIndex": "Title",
          "key": "Title",
          "title": "Title",
        },
        Object {
          "dataIndex": "Year",
          "key": "Year",
          "title": "Year",
        },
        Object {
          "key": "actions",
          "render": [Function],
          "title": "Actions",
        },
      ]
    }
    css={[Function]}
    dataSource={Array []}
    loading={false}
    onRow={[Function]}
    scroll={
      Object {
        "x": true,
      }
    }
  />
  <AboutMovie
    aboutMovie={Object {}}
    isOpenModal={false}
    toggleModal={[Function]}
  />
</Fragment>
`;
